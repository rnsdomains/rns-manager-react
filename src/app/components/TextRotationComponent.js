import React, { useEffect, useState } from 'react';
import propTypes from 'prop-types';

const TextRotationComponent = ({
  heading, language, messages, timer,
}) => {
  const [counter, setCounter] = useState(0);
  const updateCounter = () => {
    // eslint-disable-next-line react/prop-types
    const next = counter === (messages.length - 1) ? 0 : counter + 1;
    setCounter(next);
  };

  useEffect(() => {
    setTimeout(() => {
      updateCounter();
    }, 6000);
    return () => clearTimeout(timer);
  }, [counter]);

  const singleMessage = messages[counter][language];
  const contentArray = singleMessage.content.split('$');
  const contentParagraph = contentArray.length === 1
    ? singleMessage.content
    : (
      <>
        {contentArray[0]}
        <strong>{contentArray[1]}</strong>
        {contentArray[2]}
      </>
    );

  return (
    <div className="text-rotation">
      <h3 className="blue">{`${heading}...`}</h3>
      <p>
        {contentParagraph}
        {messages[counter].link !== '' && (
          <>
            <br />
            <a
              href={messages[counter].link}
              target="_blank"
              rel="noopener noreferrer"
            >
              {singleMessage.link_label}
            </a>
          </>
        )}
      </p>
    </div>
  );
};

TextRotationComponent.defaultProps = {
  timer: 1000,
};

TextRotationComponent.propTypes = {
  language: propTypes.string.isRequired,
  heading: propTypes.string.isRequired,
  messages: propTypes.arrayOf(propTypes.shape()).isRequired,
  timer: propTypes.number,
};

export default TextRotationComponent;
