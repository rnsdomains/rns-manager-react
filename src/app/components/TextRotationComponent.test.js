import React from 'react';
import { mount } from 'enzyme';
import TextRotationComponent from './TextRotationComponent';
import KeyMessages from '../../languages/key_messges.json';

describe('TextRotationComponent', () => {
  const initProps = {
    language: 'en',
    heading: 'the heading',
    timer: 10000,
    messages: [
      {
        ...KeyMessages[0],
        en: {
          content: 'hello $world$.',
          link_label: 'click',
        },
        link: 'https://developers.rsk.co/',
      },
    ],
  };

  it('renders, places correct content with bold and matches snapshot', () => {
    const component = mount(<TextRotationComponent {...initProps} />);

    expect(component.find('h3').text()).toBe('the heading...');
    // test just the <strong> part, ignore the anchor
    expect(component.find('p').at(0).html().substr(0, 32)).toBe('<p>hello <strong>world</strong>.');
    // now test the anchor
    expect(component.find('a').text()).toBe('click');
    expect(component.find('a').props().href).toBe('https://developers.rsk.co/');

    expect(component).toMatchSnapshot();
  });

  it('renders different strong text', () => {
    const localProps = {
      language: 'en',
      heading: 'the heading',
      timer: 10000,
      messages: [
        {
          ...KeyMessages[0],
          en: {
            content: 'hello, $ this is a giant$ world.',
            link_label: 'click',
          },
        },
      ],
    };

    const component = mount(<TextRotationComponent {...localProps} />);
    expect(component.find('p').at(0).html().substr(0, 50)).toBe('<p>hello, <strong> this is a giant</strong> world.');
  });

  it('renders full text and does not show link', () => {
    const localProps = {
      language: 'en',
      heading: 'the heading',
      timer: 10000,
      messages: [
        {
          en: {
            content: 'text with no link or bold',
            link_label: '',
          },
          link: '',
        },
      ],
    };

    const component = mount(<TextRotationComponent {...localProps} />);
    expect(component.find('p').html()).toBe('<p>text with no link or bold</p>');
    expect(component.find('a').length).toBe(0);
  });
});
