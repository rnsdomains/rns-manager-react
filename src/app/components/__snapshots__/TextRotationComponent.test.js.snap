// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextRotationComponent renders, places correct content with bold and matches snapshot 1`] = `
<TextRotationComponent
  heading="the heading"
  language="en"
  messages={
    Array [
      Object {
        "en": Object {
          "content": "hello $world$.",
          "link_label": "click",
        },
        "es_uy": Object {
          "content": "Puede $integrar su aplicación con RNS $utilizando las bibliotecas de código abierto.",
          "link_label": "Lea las pautas de integración",
        },
        "id": "rns-00",
        "ja": Object {
          "content": "オープンソース・ライブラリを使用して、$アプリケーションをRNSと統合$できます。",
          "link_label": "統合ガイドラインを読む",
        },
        "ko": Object {
          "content": "오픈 소스 라이브러리를 통해 $나의 애플리케이션과 RNS를 통합$할 수 있습니다.",
          "link_label": "통합 지침 읽기",
        },
        "link": "https://developers.rsk.co/",
        "pt_br": Object {
          "content": "Você pode $integrar seu aplicativo ao RNS$ usando as bibliotecas de código aberto.",
          "link_label": "Leia as diretrizes de integração",
        },
        "ru": Object {
          "content": "Вы можете $интегрировать свое приложение с RNS$, используя библиотеки с открытым исходным кодом.",
          "link_label": "Ознакомиться с рекомендациями по интеграции",
        },
        "zh_cn": Object {
          "content": "您可以$使用开源代码库来与 RNS$ 整合您的应用程序。",
          "link_label": "参阅集成指南",
        },
      },
    ]
  }
  timer={10000}
>
  <div
    className="text-rotation"
  >
    <h3
      className="blue"
    >
      the heading...
    </h3>
    <p>
      hello 
      <strong>
        world
      </strong>
      .
      <br />
      <a
        href="https://developers.rsk.co/"
        rel="noopener noreferrer"
        target="_blank"
      >
        click
      </a>
    </p>
  </div>
</TextRotationComponent>
`;
